xpystac/core.py

```diff
@@ -55,14 +55,14 @@ def _(obj: pystac.Asset, **kwargs) -> xarray.Dataset:
         default_kwargs = {"engine": "zarr", "consolidated": False, "chunks": {}}
         return xarray.open_dataset(mapper, **default_kwargs, **open_kwargs, **kwargs)
 
-    if obj.media_type == pystac.MediaType.COG:
-        _import_optional_dependency("rioxarray")
-        default_kwargs = {"engine": "rasterio"}
-    elif obj.media_type == "application/vnd+zarr":
-        _import_optional_dependency("zarr")
-        default_kwargs = {"engine": "zarr"}
-    else:
-        default_kwargs = {}
+    default_kwargs = {}
+    if open_kwargs.get("engine") is None:
+        if obj.media_type == pystac.MediaType.COG:
+            _import_optional_dependency("rioxarray")
+            default_kwargs["engine"] = "rasterio"
+        elif obj.media_type == "application/vnd+zarr":
+            _import_optional_dependency("zarr")
+            default_kwargs["engine"] = "zarr"
 
     ds = xarray.open_dataset(obj.href, **default_kwargs, **open_kwargs, **kwargs)
```

Comment:
Instead of making the changes above, we can just change this one line:

```suggestion
    ds = xarray.open_dataset(obj.href, **{**default_kwargs, **open_kwargs, **kwargs})
```

Reply:
Thanks, that's a nice and simple solution! I had to look up dictionary unpacking in https://peps.python.org/pep-0448, and check how the duplicate keys will be handled:

> The keys in a dictionary remain in a right-to-left priority order, so {**{'a': 1}, 'a': 2, **{'a': 3}} evaluates to {'a': 3}. There is no restriction on the number or position of unpackings.

So the user-set `kwargs` should override the `open_kwargs` set in the STAC metadata, which should in-turn override the `default_kwargs` set by xpystac, following the logic mentioned in https://github.com/jsignell/xpystac/issues/16#issuecomment-1477819594 :smiley: